### PR TITLE
CP Subsytem - Atomic Reference

### PR DESCRIPTION
**Pull request by mdumandag**
For the AtomicReference, implementation, code samples, documentation and
tests are added. It also includes some minor docstring fixes regarding
the lists in some places and some missing checks and documentation fixes
for the AtomicLong.

depends on #223

### CODE REVIEW

File: examples/cp/atomic_reference_example.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import hazelcast
```
Comment (puzpuzpuz) [inline examples/cp/atomic_reference_example.py:1]: nit: does it make sense to add an org web sample for AtomicReference?

File: hazelcast/proxy/cp/atomic_long.py [hunk 2]
```diff
@@ -145,13 +146,18 @@ def alter(self, function):
...
-                The function applied to the currently stored value.
+                The function that alters the currently stored value.
 
         Returns:
             hazelcast.future.Future[None]:
         """
+        check_not_none(function, "Function cannot be None")
         function_data = self._to_data(function)
         codec = atomic_long_alter_codec
+        # 1 means return the new value.
```
Comment (puzpuzpuz) [inline hazelcast/proxy/cp/atomic_long.py:157]: That's a nice addition. I had to look at the codec and server-side code to understand what `1`/`0` means. 👍

File: hazelcast/proxy/cp/atomic_long.py [hunk 3]
```diff
@@ -145,13 +146,18 @@ def alter(self, function):
 
         Args:
             function (hazelcast.serialization.api.Portable or hazelcast.serialization.api.IdentifiedDataSerializable):
-                The function applied to the currently stored value.
+                The function that alters the currently stored value.
 
         Returns:
             hazelcast.future.Future[None]:
         """
+        check_not_none(function, "Function cannot be None")
```
Comment (puzpuzpuz) [inline hazelcast/proxy/cp/atomic_long.py:154]: A potentially stupid question. Don't we need to validate number inputs in other methods, like `new_value` in `get_and_set`?

**Review by puzpuzpuz — APPROVED** [review]
LGTM. Left a few questions

File: examples/cp/atomic_reference_example.py (hunk 1 above)
Comment (mdumandag) [reply examples/cp/atomic_reference_example.py:1]: I thought about adding it but I saw that we don't have the sample for the AtomicReference in 3.12 series. So, I didn't add it but I think we should discuss the structure of the org-website samples within the team and decide on a set of samples. We can add it afterward. What do you think about it?

File: hazelcast/proxy/cp/atomic_long.py (hunk 3 above)
Comment (mdumandag) [reply hazelcast/proxy/cp/atomic_long.py:160]: In fact, it makes a lot of sense. We don't have type checks in most of the codebase, we can improve that incrementally but It makes sense to add them for the new features. I also added simple tests for invalid input cases

File: examples/cp/atomic_reference_example.py (hunk 1 above)
Comment (puzpuzpuz) [reply examples/cp/atomic_reference_example.py:1]: Yes, we can add those samples later, after an in-team discussion.

**Comment by mdumandag** [comment]
Thanks for the review Andrey

**Pull request merged**